[29기 박겸영] 로그인 및 회원가입 구현

seoltang opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [x] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 로그인 및 회원가입 모달창 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
- Login 폼, Sign-in 폼, Sign-up 폼, 비밀번호 재설정 폼 모두 `<LoginLayout>` 컴포넌트를 재사용하여 모달창으로 구현함
- 유효성 검사

1. Login 폼
- email input 입력 받고 onSubmit 시 서버에 email을 POST로 전송함.
- 입력한 email이 기존에 가입되어 있는 email인지 여부를 0 or 1로 서버에서 전달받고, 가입된 email이면(1이면) Sign-in 폼을, 가입되지 않은 email이면(0이면) Sign-up 폼을 보여줌.

2. Sign-in 폼
- email과 password를 input 입력 받고 onSubmit 시 서버에 POST로 전송함.
- 로그인 성공하면 토큰을 서버에서 받아와서 세션 스토리지에 저장함.

3. Sign-up 폼
- onInput 이벤트로 패스워드 value와 패스워드 확인 value가 같은지 확인하고, 같지 않으면 에러 메시지를 보여줌.
- onSubmit 시 email, password, first name, last name을 서버에 전송함.

4. 비밀번호 재설정 폼
- email input 입력 받음 (이전에 입력했던 value 값이 유지됨)

5. 유효성 검사
- 이메일 유효성 검사
- 비밀번호 유효성 검사: 알파벳 대문자, 소문자, 숫자, 특수문자 포함 8자리 이상
- Sign-up 폼에서 input이 하나라도 입력되지 않은 채로 submit하면 에러 메시지 보여줌.

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 여러 컴포넌트의 여러 input을 하나의 state로 관리하는 방법을 알게 되었습니다.
- 조건(fetch로 받아온 데이터)에 따라 다른 컴포넌트를 보여주는 것을 연습해 볼 수 있었습니다.
- 유효성 검사를 위해 정규표현식에 대해 찾아보았습니다.
- state를 객체로 관리할 경우 값을 꺼낼 때 구조분해할당을 하는 것이 좋다는 것을 알게 되었습니다.
- 객체에서 단축속성명을 쓸 수 있음을 알게 되었습니다.

<br />

## :: 기타 질문 및 특이 사항
- Login, Signin, Signup, ForgotPassword 컴포넌트가 각각 복잡해져서 다 따로 분리하였는데 효율적인 방식이 맞는지 궁금합니다.
- Sign-up 폼에서 setErrorState가 비동기로 작동하여 first name과 last name의 setErrorState가 완료되기 전에 submit을 실행해버리는 오류가 있습니다.

seoltang committed:
Add: Login 컴포넌트 생성

seoltang committed:
Merge branch 'master' into feature/login

seoltang committed:
Add: Login 모달창
- 로그인 및 회원가입 모달창 구현
- LoginForm, SigninForm, SignupForm, ForgotPasswordForm 컴포넌트로 분리

seoltang committed:
Add: fetch
- 로그인 및 회원가입 서버와 통신하는 fetch 함수 추가

seoltang committed:
Refactor: console.log 삭제

seoltang committed:
Refactor: console.log 삭제

seoltang commented on src/components/LoginModal/SignupForm/SignupForm.js:
```diff
@@ -0,0 +1,87 @@
+    inputState.password !== inputState.passwordConfirmation;
+
+  const submitSignup = async () => {
+    const json = await (
```
이 부분은 로그인 성공 시 서버에서 받아온 데이터(사용자 이름, 카트에 담긴 상품 개수)인데 추후에 nav바와 합치면서 보완할 예정입니다.

seoltang committed:
Merge branch 'master' into feature/login

seoltang committed:
Add: SignupForm.js: useHistory
- 회원가입 성공 시 user id를 history 객체에 쿼리스트링으로 저장
- sider 이슈 해결

201steve commented on src/components/LoginModal/LoginModal.js:
```diff
@@ -0,0 +1,113 @@
+  // 백엔드 통신용
+  // const submitLogin = async () => {
+  //   const json = await (
+  //     await fetch('http://172.30.1.11:8000/users/emailvalid', {
```
* base url

201steve commented on src/components/LoginModal/SigninForm/SigninForm.js:
```diff
@@ -0,0 +1,56 @@
+}) => {
+  const submitSignin = async () => {
+    const json = await (
+      await fetch('http://172.30.1.11:8000/users/login', {
```
* base URL

seoltang committed:
Fix: conflict resolve

seoltang committed:
Fix: conflict resolve

seoltang committed:
Merge branch 'feature/common' into feature/login

seoltang committed:
Modify: API 주소 config.js에서 import

201steve commented on src/components/LoginModal/LoginModal.js:
```diff
@@ -0,0 +1,107 @@
+    [isForgotPassword]: false,
+  });
+
+  const [inputState, setInputState] = useState({
```
* email 저장하는 state인데, 이름은 inputState여서 어울리지않습니다.
* state 변수명을 다시 생각 해 주세요

seoltang replied:
이메일뿐만 아니라 로그인 시 패스워드, 회원가입 시 패스워드와 패스워드 확인, 이름 등을 다 담는 state라서 state의 초기값을 이메일과 패스워드로 설정해 두겠습니다. 😀

201steve commented on src/components/LoginModal/SignupForm/SignupForm.js:
```diff
@@ -0,0 +1,99 @@
+  const isError =
+    inputState.password &&
+    inputState.passwordConfirmation &&
+    inputState.password !== inputState.passwordConfirmation;
```
```suggestion
const {password , passwordConfirmation} = inputState
  const isError = 
  password && 
  passwordConfirmation && 
  password !== passwordConfirmation;
```
* 이렇게 구조분해 할당하면 더 직관적으로 key만 볼 수 있습니다

seoltang committed:
Merge branch 'master' into feature/login

201steve commented on src/components/LoginModal/SignupForm/SignupForm.js:
```diff
@@ -0,0 +1,99 @@
+      await fetch(api.signUp, {
+        method: 'POST',
+        body: JSON.stringify({
+          email: inputState.email,
+          password: inputState.password,
+          'last-name': inputState.lastName,
+          'first-name': inputState.firstName,
```
위에 구조분해할당 한 inputState에 이어서, email , password도 줄이면
```suggestion
          email,
          password,
          'last-name': lastName,
          'first-name': firstName,
```

단축속성명으로 줄일 수 있습니다

201steve commented on src/components/LoginModal/SignupForm/SignupForm.js:
```diff
@@ -0,0 +1,99 @@
+    ).json();
+
+    if (json.message === 'SUCCESS') {
+      history.push({
```
* react-router-dom V6 이전에는 `history 객체`를 썻는데 
* react-router-dom V6 이후부터는 `useNavigate`를 사용합니다

seoltang committed:
Add: 유효성 검사
- SignupForm: 비밀번호 유효성 검사 및 필수항목 유효성 검사 추가
- 이메일 유효성 검사 추가
- style 수정
- ForgotPasswordForm에 onSubmit 함수 추가 (submit 시 오류 방지)

seoltang committed:
Merge branch 'master' into feature/login

seoltang committed:
Refactor: PR 리뷰 반영
- 객체 state 구조분해할당
- post 시 body 객체에 단축속성명 사용
